pkg/cloud/azure/actuators/cluster/reconciler.go

```diff
@@ -17,49 +17,259 @@ limitations under the License.
+	"sigs.k8s.io/cluster-api-provider-azure/pkg/cloud/azure/services/internalloadbalancers"
+	"sigs.k8s.io/cluster-api-provider-azure/pkg/cloud/azure/services/publicips"
+	"sigs.k8s.io/cluster-api-provider-azure/pkg/cloud/azure/services/publicloadbalancers"
+	"sigs.k8s.io/cluster-api-provider-azure/pkg/cloud/azure/services/routetables"
+	"sigs.k8s.io/cluster-api-provider-azure/pkg/cloud/azure/services/securitygroups"
+	"sigs.k8s.io/cluster-api-provider-azure/pkg/cloud/azure/services/subnets"
+	"sigs.k8s.io/cluster-api-provider-azure/pkg/cloud/azure/services/virtualnetworks"
 )
 
 // Reconciler are list of services required by cluster actuator, easy to create a fake
 type Reconciler struct {
-	scope           *actuators.Scope
-	groupsSvc       services.Service
-	certificatesSvc services.Service
+	scope            *actuators.Scope
+	certificatesSvc  services.Service
+	groupsSvc        services.Service
+	vnetSvc          services.Service
+	securityGroupSvc services.Service
+	routeTableSvc    services.Service
+	subnetsSvc       services.Service
+	internalLBSvc    services.Service
+	publicIPSvc      services.Service
+	publicLBSvc      services.Service
 }
 
 // NewReconciler populates all the services based on input scope
 func NewReconciler(scope *actuators.Scope) *Reconciler {
 	return &Reconciler{
-		scope:           scope,
-		groupsSvc:       groups.NewService(scope),
-		certificatesSvc: certificates.NewService(scope),
+		scope:            scope,
+		certificatesSvc:  certificates.NewService(scope),
+		groupsSvc:        groups.NewService(scope),
+		vnetSvc:          virtualnetworks.NewService(scope),
+		securityGroupSvc: securitygroups.NewService(scope),
+		routeTableSvc:    routetables.NewService(scope),
+		subnetsSvc:       subnets.NewService(scope),
+		internalLBSvc:    internalloadbalancers.NewService(scope),
+		publicIPSvc:      publicips.NewService(scope),
+		publicLBSvc:      publicloadbalancers.NewService(scope),
 	}
 }
 
 // Reconcile reconciles all the services in pre determined order
 func (s *Reconciler) Reconcile() error {
+	if s.scope.ClusterConfig.PublicIPName == "" {
+		h := fnv.New32a()
+		h.Write([]byte(fmt.Sprintf("%s/%s/%s", s.scope.SubscriptionID, s.scope.ClusterConfig.ResourceGroup, s.scope.Cluster.Name)))
+		s.scope.ClusterConfig.PublicIPName = strings.ToLower(services.DefaultPubicIPPrefix + fmt.Sprintf("%x", h.Sum32()))
+	}
+
 	// Store cert material in spec.
-	if err := s.certificatesSvc.CreateOrUpdate(s.scope.Context); err != nil {
+	if err := s.certificatesSvc.CreateOrUpdate(s.scope.Context, nil); err != nil {
 		return errors.Wrapf(err, "failed to CreateOrUpdate certificates for cluster %s", s.scope.Cluster.Name)
 	}
 
-	if err := s.groupsSvc.CreateOrUpdate(s.scope.Context); err != nil {
+	if err := s.groupsSvc.CreateOrUpdate(s.scope.Context, nil); err != nil {
 		return errors.Wrapf(err, "failed to CreateOrUpdate resource group for cluster %s", s.scope.Cluster.Name)
 	}
+
+	vnetSpec := virtualnetworks.Spec{
```

Comment:
can we omit these spec variables and init these inline w/ the CreateOrUpdate call?

Reply:
we should also pass these as ref as well